File: server/analyzer/risk_detector.go
package analyzer

import (
	"fmt"
	"regexp"
	"strings"
)

// RiskDetector 風險檢測引擎
type RiskDetector struct {
	parseResult *ParseResult
}

// RiskItem 風險項目
type RiskItem struct {
	RiskType      string // "magic_value", "multiple_writes", "missing_condition", etc.
	Severity      string // "low", "medium", "high", "critical"
	Title         string
	Description   string
	SourceFile    string
	LineNumber    int
	CodeSnippet   string
	Recommendation string
}

// NewRiskDetector 建立新的風險檢測器
func NewRiskDetector(parseResult *ParseResult) *RiskDetector {
	return &RiskDetector{
		parseResult: parseResult,
	}
}

// Detect 執行風險檢測
func (rd *RiskDetector) Detect() []*RiskItem {
	risks := make([]*RiskItem, 0)

	// 檢測魔法值
	risks = append(risks, rd.detectMagicValues()...)

	// 檢測多處寫入同欄位
	risks = append(risks, rd.detectMultipleWrites()...)

	// 檢測缺少條件的 SQL
	risks = append(risks, rd.detectMissingConditions()...)

	// 檢測日期/金額格式轉換
	risks = append(risks, rd.detectFormatConversions()...)

	// 檢測不一致的邏輯
	risks = append(risks, rd.detectInconsistentLogic()...)

	return risks
}

// detectMagicValues 檢測魔法值
func (rd *RiskDetector) detectMagicValues() []*RiskItem {
	risks := make([]*RiskItem, 0)

	for _, mv := range rd.parseResult.MagicValues {
		// 檢查是否是常見的危險魔法值
		if isRiskyMagicValue(mv.Value) {
			risk := &RiskItem{
				RiskType:   "magic_value",
				Severity:   "medium",
				Title:      fmt.Sprintf("檢測到魔法值: %s", mv.Value),
				Description: fmt.Sprintf("在 %s 中發現魔法值 '%s' (%s)", mv.Location.File, mv.Value, mv.Context),
				SourceFile: mv.Location.File,
				LineNumber: mv.Location.Line,
				CodeSnippet: mv.Context,
				Recommendation: "建議將魔法值提取為具有明確名稱的常數，以提高程式碼可讀性和可維護性。",
			}
			risks = append(risks, risk)
		}
	}

	return risks
}

// detectMultipleWrites 檢測多處寫入同欄位
func (rd *RiskDetector) detectMultipleWrites() []*RiskItem {
	risks := make([]*RiskItem, 0)

	// 統計每個欄位的寫入次數
	writeMap := make(map[string][]*FieldReference)
	for _, ref := range rd.parseResult.FieldReferences {
		if ref.OperationType == "write" {
			key := fmt.Sprintf("%s.%s", ref.TableName, ref.FieldName)
			writeMap[key] = append(writeMap[key], ref)
		}
	}

	// 找出多處寫入的欄位
	for key, refs := range writeMap {
		if len(refs) > 2 {
			parts := strings.Split(key, ".")
			risk := &RiskItem{
				RiskType:   "multiple_writes",
				Severity:   "high",
				Title:      fmt.Sprintf("欄位 %s 有多處寫入", key),
				Description: fmt.Sprintf("欄位 '%s' 在 %d 個不同位置被寫入，可能導致資料一致性問題", key, len(refs)),
				SourceFile: refs[0].Location.File,
				LineNumber: refs[0].Location.Line,
				CodeSnippet: refs[0].Context,
				Recommendation: fmt.Sprintf("建議檢查所有 %d 個寫入位置，確保資料一致性。考慮集中管理此欄位的更新邏輯。", len(refs)),
			}
			risks = append(risks, risk)
		}
	}

	return risks
}

// detectMissingConditions 檢測缺少條件的 SQL
func (rd *RiskDetector) detectMissingConditions() []*RiskItem {
	risks := make([]*RiskItem, 0)

	// 尋找 UPDATE/DELETE 語句但沒有 WHERE 條件
	updateDeleteRegex := regexp.MustCompile(`(?i)(UPDATE|DELETE)\s+(\w+)(?:\s+SET)?(?:\s+[^;]*)?(?:;|$)`)

	for _, symbol := range rd.parseResult.Symbols {
		if symbol.Type == "query" {
			if matches := updateDeleteRegex.FindStringSubmatch(symbol.Content); matches != nil {
				operation := matches[1]
				tableName := matches[2]

				// 檢查是否有 WHERE 條件
				if !strings.Contains(strings.ToUpper(symbol.Content), "WHERE") {
					risk := &RiskItem{
						RiskType:   "missing_condition",
						Severity:   "critical",
						Title:      fmt.Sprintf("%s 語句缺少 WHERE 條件", operation),
						Description: fmt.Sprintf("在 %s:%d 發現 %s 語句操作表 '%s' 但缺少 WHERE 條件，可能影響所有記錄", symbol.Location.File, symbol.Location.Line, operation, tableName),
						SourceFile: symbol.Location.File,
						LineNumber: symbol.Location.Line,
						CodeSnippet: symbol.Content,
						Recommendation: "務必添加 WHERE 條件以限制操作範圍。如果確實需要操作所有記錄，應明確確認並添加註解說明。",
					}
					risks = append(risks, risk)
				}
			}
		}
	}

	return risks
}

// detectFormatConversions 檢測日期/金額格式轉換
func (rd *RiskDetector) detectFormatConversions() []*RiskItem {
	risks := make([]*RiskItem, 0)

	// 尋找日期格式轉換的模式
	dateFormatRegex := regexp.MustCompile(`(?i)(strtotime|date|time\.Parse|format|strftime)\s*\(`)
	amountFormatRegex := regexp.MustCompile(`(?i)(tostring|tostring|parseFloat|parseDecimal|round)\s*\(`)

	for _, symbol := range rd.parseResult.Symbols {
		// 檢查日期格式轉換
		if dateFormatRegex.MatchString(symbol.Content) {
			// 檢查是否有補零邏輯
			if !strings.Contains(symbol.Content, "pad") && !strings.Contains(symbol.Content, "zero") {
				risk := &RiskItem{
					RiskType:   "format_conversion",
					Severity:   "medium",
					Title:      "日期格式轉換可能缺少補零邏輯",
					Description: fmt.Sprintf("在 %s:%d 發現日期格式轉換，但未見補零邏輯", symbol.Location.File, symbol.Location.Line),
					SourceFile: symbol.Location.File,
					LineNumber: symbol.Location.Line,
					CodeSnippet: symbol.Content,
					Recommendation: "確保日期格式轉換時使用正確的補零邏輯（例如 yyyyMMdd 格式需要補零）。",
				}
				risks = append(risks, risk)
			}
		}

		// 檢查金額格式轉換
		if amountFormatRegex.MatchString(symbol.Content) {
			// 檢查是否有小數位處理
			if !strings.Contains(symbol.Content, "decimal") && !strings.Contains(symbol.Content, "precision") {
				risk := &RiskItem{
					RiskType:   "format_conversion",
					Severity:   "high",
					Title:      "金額格式轉換可能缺少小數位處理",
					Description: fmt.Sprintf("在 %s:%d 發現金額格式轉換，但未見小數位處理邏輯", symbol.Location.File, symbol.Location.Line),
					SourceFile: symbol.Location.File,
					LineNumber: symbol.Location.Line,
					CodeSnippet: symbol.Content,
					Recommendation: "確保金額轉換時正確處理小數位。建議使用 Decimal 類型而非 Float 以避免精度問題。",
				}
				risks = append(risks, risk)
			}
		}
	}

	return risks
}

// detectInconsistentLogic 檢測不一致的邏輯
func (rd *RiskDetector) detectInconsistentLogic() []*RiskItem {
	risks := make([]*RiskItem, 0)

	// 尋找相同的欄位被以不同方式處理的情況
	fieldHandlingMap := make(map[string]map[string]int)

	for _, ref := range rd.parseResult.FieldReferences {
		key := fmt.Sprintf("%s.%s", ref.TableName, ref.FieldName)
		if fieldHandlingMap[key] == nil {
			fieldHandlingMap[key] = make(map[string]int)
		}
		fieldHandlingMap[key][ref.Context]++
	}

	// 檢查是否有不一致的處理
	for key, contexts := range fieldHandlingMap {
		if len(contexts) > 2 {
			risk := &RiskItem{
				RiskType:   "inconsistent_logic",
				Severity:   "medium",
				Title:      fmt.Sprintf("欄位 %s 有不一致的處理邏輯", key),
				Description: fmt.Sprintf("欄位 '%s' 在不同位置以不同方式被處理，可能導致邏輯不一致", key),
				Recommendation: "建議檢查所有對此欄位的處理邏輯，確保一致性。考慮統一處理方式。",
			}
			risks = append(risks, risk)
		}
	}

	return risks
}

// isRiskyMagicValue 判斷是否是危險的魔法值
func isRiskyMagicValue(value string) bool {
	riskyPatterns := []string{
		"^[YN]$",           // Y/N 標記
		"^[0-9]{2}$",       // 兩位數代碼
		"^[0-9]{4}$",       // 四位數代碼
		"^00$", "^01$", "^99$", // 特殊代碼
	}

	for _, pattern := range riskyPatterns {
		if matched, _ := regexp.MatchString(pattern, value); matched {
			return true
		}
	}

	return false
}

File: server/analyzer/riskDetector.ts
import type { DetectedRisk, FieldReference } from "./types";

export interface MagicValueCandidate {
  value: string;
  file: string;
  line: number;
  context: string;
  kind: "string" | "number";
}

export class RiskDetector {
  detectMagicValues(candidates: MagicValueCandidate[]): DetectedRisk[] {
    return candidates.flatMap((candidate) => {
      if (["0", "1", "-1", "100"].includes(candidate.value)) {
        return [];
      }

      const risks: DetectedRisk[] = [];
      const isDateLike = /^\d{8}$/.test(candidate.value) || /^\d{4}-\d{2}-\d{2}$/.test(candidate.value);

      if (isDateLike) {
        risks.push({
          title: "Date literal embedded in code",
          description: `Found hard-coded date value "${candidate.value}" in executable code.`,
          severity: "medium",
          category: "magic_value",
          sourceFile: candidate.file,
          lineNumber: candidate.line,
          suggestion: "Promote the value to a named constant or configuration entry.",
          codeSnippet: candidate.context,
        });
      } else if (candidate.kind === "number" && candidate.value.length >= 4) {
        risks.push({
          title: "Large numeric literal embedded in code",
          description: `Found numeric literal "${candidate.value}" that should be documented or externalized.`,
          severity: "medium",
          category: "magic_value",
          sourceFile: candidate.file,
          lineNumber: candidate.line,
          suggestion: "Replace the literal with a named constant and document the business meaning.",
          codeSnippet: candidate.context,
        });
      } else if (candidate.kind === "string" && candidate.value.length >= 10) {
        risks.push({
          title: "String literal with business meaning",
          description: `Found long string literal "${candidate.value.slice(0, 40)}" embedded in code.`,
          severity: "low",
          category: "magic_value",
          sourceFile: candidate.file,
          lineNumber: candidate.line,
          suggestion: "Consider centralizing repeated business literals.",
          codeSnippet: candidate.context,
        });
      }

      return risks;
    });
  }

  detectMultipleWrites(fieldReferences: FieldReference[]): DetectedRisk[] {
    const writeMap = new Map<string, FieldReference[]>();

    for (const reference of fieldReferences) {
      if (reference.type !== "write") continue;
      const key = `${reference.table}.${reference.field}`;
      const bucket = writeMap.get(key) ?? [];
      bucket.push(reference);
      writeMap.set(key, bucket);
    }

    const risks: DetectedRisk[] = [];
    for (const [fieldKey, references] of Array.from(writeMap.entries())) {
      if (references.length < 2) continue;
      const first = references[0];
      risks.push({
        title: `Field ${fieldKey} is written from multiple locations`,
        description: `${fieldKey} has ${references.length} write operations across the codebase, which raises drift risk.`,
        severity: references.length >= 3 ? "high" : "medium",
        category: "multiple_writes",
        sourceFile: first.file,
        lineNumber: first.line,
        suggestion: "Review whether the field should have a single write owner or stronger validation.",
        codeSnippet: references.map((reference: FieldReference) => `${reference.file}:${reference.line}`).join(", "),
      });
    }

    return risks;
  }

  detectMissingConditions(sqlSnippets: Array<{ file: string; line: number; sql: string }>): DetectedRisk[] {
    return sqlSnippets.flatMap((snippet) => {
      if (!/^(UPDATE|DELETE)\b/i.test(snippet.sql)) {
        return [];
      }

      if (/\bWHERE\b/i.test(snippet.sql)) {
        return [];
      }

      return [
        {
          title: "Write statement without WHERE clause",
          description: "Detected UPDATE/DELETE statement without a WHERE clause.",
          severity: "critical",
          category: "missing_condition",
          sourceFile: snippet.file,
          lineNumber: snippet.line,
          suggestion: "Add an explicit predicate or document why full-table mutation is safe.",
          codeSnippet: snippet.sql,
        },
      ];
    });
  }

  detectDelphiPatterns(content: string, file: string): DetectedRisk[] {
    const risks: DetectedRisk[] = [];
    const lines = content.split(/\r?\n/);
    const seen = new Set<string>();
    let i = 0;

    while (i < lines.length) {
      const line = lines[i] ?? "";
      const trimmed = line.trim();

      if (!trimmed || trimmed.startsWith("//") || trimmed.startsWith("#") || trimmed.startsWith("{") || trimmed.startsWith("(*")) {
        i++;
        continue;
      }

      const fieldByNameMatch = trimmed.match(/FieldByName\(\s*['"]([^'"]+)['"]\s*\)/i);
      if (fieldByNameMatch) {
        const key = `FieldByName:${fieldByNameMatch[1]}:${i}`;
        if (!seen.has(key)) {
          seen.add(key);
          risks.push({
            title: "Delphi field access detected",
            description: `Found FieldByName("${fieldByNameMatch[1]}") usage. Review data access patterns and result validation.`,
            severity: "low",
            category: "other",
            sourceFile: file,
            lineNumber: i + 1,
            suggestion: "Confirm that field names are valid and that the data access is protected against missing fields.",
            codeSnippet: trimmed,
          });
        }
      }

      const paramByNameMatch = trimmed.match(/ParamByName\(\s*['"]([^'"]+)['"]\s*\)/i);
      if (paramByNameMatch) {
        const key = `ParamByName:${paramByNameMatch[1]}:${i}`;
        if (!seen.has(key)) {
          seen.add(key);
          risks.push({
            title: "Delphi query parameter usage detected",
            description: `Found ParamByName("${paramByNameMatch[1]}"); verify parameters are used consistently with the SQL statement.`,
            severity: "low",
            category: "other",
            sourceFile: file,
            lineNumber: i + 1,
            suggestion: "Verify that parameters are applied to parameterized SQL statements rather than string-concatenated queries.",
            codeSnippet: trimmed,
          });
        }
      }

      if (/\bSQL\.Add\s*\(/i.test(trimmed) && /\+/.test(trimmed)) {
        const key = `SQL.Add.dynamic:${i}`;
        if (!seen.has(key)) {
          seen.add(key);
          risks.push({
            title: "Dynamic Delphi SQL construction",
            description: "Detected SQL.Add with string concatenation; this can expose the application to SQL injection or broken queries.",
            severity: "high",
            category: "other",
            sourceFile: file,
            lineNumber: i + 1,
            suggestion: "Use parameterized query building and avoid concatenating user input into SQL strings.",
            codeSnippet: trimmed,
          });
        }
      }

      if (/\bSQL\.Text\b.*:=.*\+/.test(trimmed)) {
        const key = `SQL.Text.dynamic:${i}`;
        if (!seen.has(key)) {
          seen.add(key);
          risks.push({
            title: "Dynamic SQL text assignment",
            description: "Detected SQL.Text assignment using string concatenation; this is a common source of query risk.",
            severity: "high",
            category: "other",
            sourceFile: file,
            lineNumber: i + 1,
            suggestion: "Review the SQL text and convert it to parameterized queries where possible.",
            codeSnippet: trimmed,
          });
        }
      }

      if (/\bexcept\b/i.test(trimmed) && !/raise\b/i.test(trimmed)) {
        let j = i + 1;
        let foundEmptyHandler = false;
        while (j < Math.min(i + 10, lines.length)) {
          const nextTrimmed = (lines[j] ?? "").trim();
          if (!nextTrimmed) {
            j++;
            continue;
          }
          if (nextTrimmed === ";") {
            foundEmptyHandler = true;
            break;
          }
          if (/^end;?$/i.test(nextTrimmed) && !foundEmptyHandler) {
            foundEmptyHandler = true;
            break;
          }
          if (/\b(on|raise|begin|try)\b/i.test(nextTrimmed) || /^[^;]*:=/.test(nextTrimmed)) {
            break;
          }
          j++;
        }
        if (foundEmptyHandler) {
          const key = `empty.except:${i}`;
          if (!seen.has(key)) {
            seen.add(key);
            risks.push({
              title: "Broad or empty Delphi exception handling",
              description: "Found an except block without a concrete handling action; this can suppress runtime failures.",
              severity: "high",
              category: "other",
              sourceFile: file,
              lineNumber: i + 1,
              suggestion: "Avoid empty exception handlers and handle or log exceptions explicitly.",
              codeSnippet: trimmed,
            });
          }
        }
      }

      const dbComponentMatch = trimmed.match(/\b(TQuery|TADOQuery|TClientDataSet|TFDQuery|DataSource)\b/i);
      if (dbComponentMatch) {
        const key = `query.type:${i}`;
        if (!seen.has(key)) {
          seen.add(key);
          risks.push({
            title: "Delphi database component detected",
            description: `Found Delphi database component usage (${dbComponentMatch[1]}). Review connection and query handling patterns.`,
            severity: "low",
            category: "other",
            sourceFile: file,
            lineNumber: i + 1,
            suggestion: "Check whether the component uses parameterized queries and proper transaction boundaries.",
            codeSnippet: trimmed,
          });
        }
      }

      if (/([A-Za-z]:\\|\\\\)[^"'\s]+/.test(trimmed) || /\b(Database|Data Source|User ID|Password|Pwd|Server|Provider)\b/i.test(trimmed)) {
        const key = `hardcoded.path:${i}`;
        if (!seen.has(key)) {
          seen.add(key);
          risks.push({
            title: "Hardcoded Delphi path or connection string",
            description: "Detected a hardcoded filesystem path or connection string pattern in Delphi code.",
            severity: "medium",
            category: "other",
            sourceFile: file,
            lineNumber: i + 1,
            suggestion: "Move environment-specific values to configuration and avoid literal paths in code.",
            codeSnippet: trimmed,
          });
        }
      }

      i++;
    }

    return risks;
  }

  detectFormatConversionRisks(content: string, file: string): DetectedRisk[] {
    const lines = content.split(/\r?\n/);
    const patterns: Array<{ regex: RegExp; severity: DetectedRisk["severity"]; label: string }> = [
      { regex: /Parse(Time|Date)|Format(Date|Time)|strftime/i, severity: "medium", label: "date/time conversion" },
      { regex: /ParseFloat|Round\(|Truncate\(|Decimal/i, severity: "high", label: "amount conversion" },
    ];

    const risks: DetectedRisk[] = [];
    lines.forEach((line, index) => {
      for (const pattern of patterns) {
        if (!pattern.regex.test(line)) {
          continue;
        }

        risks.push({
          title: `Potential ${pattern.label} inconsistency`,
          description: `Found ${pattern.label} logic that should be aligned with documented business rules.`,
          severity: pattern.severity,
          category: "format_conversion",
          sourceFile: file,
          lineNumber: index + 1,
          suggestion: "Verify precision, timezone, and canonical format rules against the target system.",
          codeSnippet: line.trim(),
        });
      }
    });

    return risks;
  }

  calculateRiskScore(risks: DetectedRisk[]): number {
    const weights: Record<DetectedRisk["severity"], number> = {
      critical: 40,
      high: 20,
      medium: 8,
      low: 3,
    };

    return risks.reduce((total, risk) => total + weights[risk.severity], 0);
  }
}